lib/Sema/CSGen.cpp

```diff
@@ -1003,8 +1003,17 @@ namespace {
     ConstraintSystem &getConstraintSystem() const { return CS; }
 
     virtual Type visitErrorExpr(ErrorExpr *E) {
-      // FIXME: Can we do anything with error expressions at this point?
-      return nullptr;
+      if (!CS.isForCodeCompletion())
+        return nullptr;
+
+      // For code completion, treat error expressions that don't contain
+      // the completion location itself as holes.
+      SourceRange range = E->getSourceRange();
+      if (range.isInvalid() ||
```

Comment:
Nit: Could you please leave a small comment explaining why we don't want to do this when there is an "original" expression inside of `ErrorExpr`?

Reply:
If the completion position is inside `ErrorExpr`, could we walk into `OriginalExpr` to generate constraints for them? So that we could get rid of the fallback typechecking? I'm not sure that is a good solution, but just curios.

Reply:
I think we'll still need the fallback regardless because of multi-statement closure bodies being checked in a separate call to typeCheckExpression than the outer expression containing them. If the completion is in a multi-statement closure body and the outer typeCheckExpression call produces a solution, we wait for the second typeCheckExpression call for the closure body expressions to determine the base and context types for completion. If the outer typeCheckExpression call fails to produce a solution there is no second call for the expressions in the multi-statement closure body it contained so we have to do a fallback typecheck on the expression containing the completion within the body (and we won't know the closure's type unless it was explicitly written since the outer type check failed).

@xedin If we walk into the original expression would the error expression's original expression still solve in isolation from its surrounding context? I.e in `x = foo(someVar.someMember.).<here>` the argument to foo is an error expression because of the trailing `.` before the `)` with no following member. if we walk the `someVar.someMember` part during constraint generation would we end up trying to match someMember's type against the parameter type of any foo overloads?

Reply:
> If we walk into the original expression would the error expression's original expression still solve in isolation from its surrounding context? 

No, it would become a part of a larger constraint system.